community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexProviderTest.java

```diff
@@ -0,0 +1,422 @@
+
+    private static final int indexId = 1;
+    private static final int labelId = 1;
+    private static final int propId = 1;
+    private SchemaIndexProvider provider;
+    private final AssertableLogProvider logging = new AssertableLogProvider();
+    private final SchemaIndexProvider.Monitor monitor = new LoggingMonitor( logging.getLog( "test" ) );
+
+    @Before
+    public void setup() throws IOException
+    {
+        File nativeSchemaIndexStoreDirectory = newProvider().directoryStructure().rootDirectory();
+        rules.fileSystem().mkdirs( nativeSchemaIndexStoreDirectory );
+    }
+
+    /* getPopulator */
+
+    @Test
+    public void getPopulatorMustThrowIfInReadOnlyMode() throws Exception
+    {
+        // given
+        provider = newReadOnlyProvider();
+
+        try
+        {
+            // when
+            provider.getPopulator( indexId, descriptor(), samplingConfig() );
+            fail( "Should have failed" );
+        }
+        catch ( UnsupportedOperationException e )
+        {
+            // then
+            // good
+        }
+    }
+
+    @Test
+    public void getPopulatorMustCreateUniquePopulatorForTypeUnique() throws Exception
+    {
+        // given
+        provider = newProvider();
+
+        // when
+        IndexPopulator populator = provider.getPopulator( indexId, descriptorUnique(), samplingConfig() );
+
+        // then
+        assertTrue( "Expected populator to be unique populator", populator instanceof NativeUniqueSchemaIndexPopulator );
+    }
+
+    @Test
+    public void getPopulatorMustCreateNonUniquePopulatorForTypeGeneral() throws Exception
+    {
+        // given
+        provider = newProvider();
+
+        // when
+        IndexPopulator populator = provider.getPopulator( indexId, descriptor(), samplingConfig() );
+
+        // then
+        assertTrue( "Expected populator to be non-unique populator", populator instanceof NativeNonUniqueSchemaIndexPopulator );
+    }
+
+    /* getOnlineAccessor */
+
+    @Test
+    public void getOnlineAccessorMustCreateUniqueAccessorForTypeUnique() throws Exception
+    {
+        // given
+        provider = newProvider();
+
+        // when
+        IndexDescriptor descriptor = descriptorUnique();
+        try ( IndexAccessor accessor = provider.getOnlineAccessor( indexId, descriptor, samplingConfig() );
+              IndexUpdater indexUpdater = accessor.newUpdater( IndexUpdateMode.ONLINE ) )
+        {
+            Value value = someValue();
+            indexUpdater.process( IndexEntryUpdate.add( 1, descriptor.schema(), value ) );
+
+            // then
+            try
+            {
+                indexUpdater.process( IndexEntryUpdate.add( 2, descriptor.schema(), value ) );
+                fail( "Should have failed" );
+            }
+            catch ( IndexEntryConflictException e )
+            {
+                // good
+            }
+        }
+    }
+
+    protected abstract Value someValue();
```

Comment:
nice to see this hidden abstract method between the tests, otherwise its to easy to find all f those abstract things...